components/job-orchestration/job_orchestration/scheduler/search/search_scheduler.py

```diff
@@ -210,32 +250,55 @@ def check_job_status_and_update_db(db_conn):
             continue
 
         if returned_results is not None:
-            new_job_status = SearchJobStatus.SUCCEEDED
+            new_job_status = SearchJobStatus.RUNNING
             for task_result_obj in returned_results:
                 task_result = SearchTaskResult.parse_obj(task_result_obj)
                 if not task_result.success:
                     task_id = task_result.task_id
                     new_job_status = SearchJobStatus.FAILED
                     logger.debug(f"Task {task_id} failed - result {task_result}.")
 
-            del active_jobs[job_id]
-
-            if set_job_status(db_conn, job_id, new_job_status, SearchJobStatus.RUNNING):
-                if new_job_status != SearchJobStatus.FAILED:
-                    logger.info(f"Completed job {job_id}.")
-                else:
-                    logger.info(f"Completed job {job_id} with failing tasks.")
+            if new_job_status != SearchJobStatus.FAILED:
+                max_num_results = job.search_config.max_num_results
+                # Check if we've searched all archives
+                if len(job.remaining_archives_for_search) == 0:
+                    new_job_status = SearchJobStatus.SUCCEEDED
+                # Check if we've reached max results
+                elif max_num_results > 0:
+                    results_cache_client = pymongo.MongoClient(results_cache_uri)
+                    results_cache_collection = results_cache_client.get_default_database()[job_id]
+                    results_count = results_cache_collection.count_documents({})
+                    if results_count >= max_num_results:
+                        max_timestamp_in_archive = job.remaining_archives_for_search[0][
+                            "end_timestamp"
+                        ]
+                        if max_timestamp_in_archive <= get_min_timestamp_in_top_results(
+                            results_cache_collection, max_num_results
+                        ):
+                            new_job_status = SearchJobStatus.SUCCEEDED
+                    results_cache_client.close()
```

Comment:
Can we also put the connection opening and closing in the method? Since we're not reusing the connection besides for the purpose of getting the min timestamp in the top results.

Reply:
But we also need to count the documents, if `result_count < max_num_results`, we don't need to `get_min_timestamp_in_top_results`

Reply:
Does this method include both counting the documents and getting min timestamp in the top results? And what's the return value?

Reply:
Something like this:

```py
def found_max_num_latest_results(
    results_cache_uri: str,
    job_id: str,
    max_num_results: int,
    max_timestamp_in_remaining_archives: int,
) -> bool:
    with pymongo.MongoClient(results_cache_uri) as results_cache_client:
        results_cache_collection = results_cache_client.get_default_database()[job_id]
        results_count = results_cache_collection.count_documents({})
        if results_count < max_num_results:
            return False

        results = list(
            results_cache_collection.find(
                projection=["timestamp"],
                sort=[("timestamp", pymongo.DESCENDING)],
                limit=max_num_results,
            )
            .sort("timestamp", pymongo.ASCENDING)
            .limit(1)
        )
        min_timestamp_in_top_results = 0 if len(results) == 0 else results[0]["timestamp"]
        return max_timestamp_in_remaining_archives <= min_timestamp_in_top_results
```

Note some of the query optimizations.